crates/admin/src/cluster_controller/observed_cluster_state.rs

```diff
@@ -38,6 +38,20 @@ impl ObservedClusterState {
     }
 
     pub fn update(&mut self, cluster_state: &ClusterState) {
+        // In case nodes were removed and no longer exists in the cluster_state;
+        // we make sure they are completely removed from both dead and alive sets.
+        self.alive_nodes
+            .retain(|id, _| cluster_state.nodes.contains_key(id));
+        self.dead_nodes
+            .retain(|id| cluster_state.nodes.contains_key(id));
+        self.nodes_to_partitions
+            .retain(|id, _| cluster_state.nodes.contains_key(id));
+        for partition_state in self.partitions.values_mut() {
+            partition_state
+                .partition_processors
+                .retain(|id, _| cluster_state.nodes.contains_key(id));
+        }
```

Comment:
I think this is correct. The one thing that bothers me a little bit is that in most cases we will retain all the information. So basically burning a bunch of CPU cycles for little gain. If we cleaned up the `ObservedClusterState` struct on a `NodesConfiguration` change, then we have the problem that an `update` call might come from before the `NodesConfiguration` has been changed and then re-insert a removed node.

Probably this is premature optimization since we are doing a couple of processing steps on the cluster state update anyway and this just adds a little bit more. Maybe something to measure and improve in the future.

Reply:
@tillrohrmann  Yeah, I totally get your concern. I first wanted to do it inline (during the update) but i ended up with similar iterations to remove any node that does not exist anymore, which is not more performant than the `retains`. Another solution I considered is to make sure all tombstone nodes are included in the cluster_state as `Dead` but I thought it might be misleading and or have other side effects that I might not noticed.